src/tir/transforms/lower_sparse_tir.cc

```diff
@@ -281,43 +283,120 @@ class IndexTransformer : public StmtExprMutator {
+    do {
+      /* Block itervars of current block. */
+      Array<IterVar> block_iters;
+      /* Itervar bindings of current block. */
+      Array<PrimExpr> iter_bindings;
+      /* Axis names of current block. */
+      Array<String> axis_names;
+      /* Genereated loop vars of current block. */
+      Array<Var> loop_vars;
+      /* An indicator that records whether there is reduction axis in current block. */
+      bool has_reduction_var = false;
+      for (int i = 0; i < n_iter; ++i) {
+        SpIterVar sp_it_var = sp_block->sp_iter_vars[i];
+        String axis_name = sp_it_var->axis->name;
+        auto&& parent_axis = t_->parent.Get(axis_name);
+        CHECK(parent_axis.defined()) << "Sparse IterVar not defined in Axis Tree.";
+        String parent_axis_name = parent_axis.value();
+        bool is_fixed_axis = sp_it_var->axis->is_fixed();
+        /* Add itervar to current block when
+         * - it's not used yet (not in stack) and
+         *   - it's parent axis was used in outer blocks or
+         *   - it's an iterator to a fixed axis.
+         */
+        if ((is_fixed_axis || in_stack.find(parent_axis_name) != in_stack.end()) &&
+            in_stack.find(axis_name) == in_stack.end()) {
+          loop_vars.push_back(all_loop_vars[i]);
+          axis_names.push_back(std::move(axis_name));
+          block_iters.push_back(SpIterVarToIterVar(sp_it_var, var_map));
+          iter_bindings.push_back(all_loop_vars[i]);
+          has_reduction_var |= sp_it_var->is_reduction;
+        }
+      }
+
+      /* Tag axes in current block as "in-stack". */
+      for (const String&& axis_name : axis_names) {
+        in_stack.insert(std::move(axis_name));
+      }
+
+      /* Update stack. */
+      if (!block_iters.empty()) {
+        block_iters_st.push(std::move(block_iters));
+        iter_bindings_st.push(std::move(iter_bindings));
+        loop_vars_st.push(std::move(loop_vars));
+        if (init_set) {
+          place_init_st.push(false);
+        } else {
+          place_init_st.push(has_reduction_var);
+          init_set |= has_reduction_var;
+        }
+      } else {
+        break;
+      }
+    } while (true);
 
     // Step 4. Generate the read-region and write-retion of the block.
     Array<BufferRegion> reads{nullptr};
     Array<BufferRegion> writes{nullptr};
     GenerateReadWriteRegions(sp_block, &reads, &writes);
 
-    // Step 5. Create the block and block-realize
-    Map<String, ObjectRef> mapping;
-    mapping.Set("sparse", Bool(true));
-    Block block(block_iters, std::move(reads), std::move(writes), sp_block->name, std::move(body),
-                std::move(init), {}, {}, std::move(mapping));
-    BlockRealize block_realize(std::move(iter_bindings), const_true(), std::move(block));
-
-    // Step 6. Create outer loops and the block binding.
-    Stmt loop = GenerateLoops(std::move(block_realize), block_iters, loop_vars);
+    // Step 5. Generate nested blocks and loops from innermost to outermost.
```

Comment:
Haven't got enough time to review yet. I'm going to think more on this part tomorrow.